docs(Navigation): add accessibility documentation (#3254)
patternfly/patternfly-org: packages/v4/patternfly-docs/content/accessibility/navigation/navigation.md

---
id: Navigation
section: components
---

import { Checkbox, List, ListItem } from '@patternfly/react-core';

## Accessibility

To implement an accessible PatternFly **navigation** component:
- Provide an accessible name for the navigation via the `aria-label` attribute
- For grouped or expandable nav, add a `title` to each section to distinguish them for users. If `title` is not used, an `aria-label` must be used on the `NavGroup` or `NavExpandable` component.

## Testing

At a minimum, a navigation component should meet the following criteria:

<List isPlain>
  <ListItem>
    <Checkbox id="navigation-a11y-checkbox-1" label={<span>The navigation has an accessible name via the <code class="ws-code">aria-label</code></span>} description="This will differentiate each navigation when announced to users of assistive technologies when navigating through a page or a screen reader's rotor menu." />
  </ListItem>
  <ListItem>

>     <Checkbox id="navigation-a11y-checkbox-2" label="Standard keyboard navigation can be used to navigate between focusable elements within the navigation." description={<span><kbd>Tab</kbd> navigates to the next focusable element, and <kbd>Shift</kbd> + <kbd>Tab</kbd> navigates to the previous focusable element. Expandable items can be opened with <kbd>Enter</kbd> or <kbd>Space</kbd>. Flyout menu items use arrow keys to navigate between flyout items, and press <kbd>Escape</kbd> or left arrow to close a flyout. When the navigation is collapsed, you should not be able to navigate to any of the elements.</span>} />

Author: Ok, I added a section for flyout. I didn't for drilldown yet because I noticed that it seems to be broken at the moment. 🙃 Should I just combine the sections or not mention it at all yet or another option?

>     <Checkbox id="navigation-a11y-checkbox-2" label="Standard keyboard navigation can be used to navigate between focusable elements within the navigation." description={<span><kbd>Tab</kbd> navigates to the next focusable element, and <kbd>Shift</kbd> + <kbd>Tab</kbd> navigates to the previous focusable element. Expandable items can be opened with <kbd>Enter</kbd> or <kbd>Space</kbd>. Flyout menu items use arrow keys to navigate between flyout items, and press <kbd>Escape</kbd> or left arrow to close a flyout. When the navigation is collapsed, you should not be able to navigate to any of the elements.</span>} />

Reviewer: Ah is it the issue where the menu height shrinks and nothing inside it is actually visible? There was a similar issue with the Menu drilldown, I can take a look to see if its the same issue here.

We could try using the Menu Drilldown to base the verbiage in this PR off of, I imagine the keyboard interaction would be essentially similar. Though it could also be an easy update to make after the fact once the Nav Drilldown example is working properly just to ensure the interaction is accurate.

</List>

## React customization

Various React props have been provided for more fine-tuned control over accessibility.

| Prop | Applied to | Reason | 
|---|---|---|
| `aria-label` | `Nav` | 	Accessible label for the nav when there are multiple navs on the page |
| `ariaLeftScroll` | `NavList` | Aria-label for the left scroll button |
| `ariaRightScroll` | `NavList` | Aria-label for the right scroll button |
| `title` | `NavExpandable` | Title shown for the expandable list |
| `buttonProps` | `NavExpandable` | Additional props added to the NavExpandable `<button>`|
| `srText` | `NavExpandable` | If defined, screen readers will read this text instead of the list title |

### Expandable toggle props

When a navigation has expandable content, `buttonProps` must be passed into the `NavExpandable` sub-component with an object as its value. This object should usually be passed in with a structure similar to the following:

```noLive
  buttonProps=({
    id: ‘buttonId’, 
    ‘aria-label’: 'Toggle label', 
    ‘aria-labelledby’: 'titleId buttonId', 
    ‘aria-expanded’: this.state.isExpanded
  )}
  ```

## HTML/CSS customization

| Attribute | Applied to | Outcome |
| -- | -- | -- |

> | `aria-label="[landmark description]"` | `.pf-c-nav` |  Describes `<nav>` landmark. |

Reviewer 2: FWIW I like having the information about what the value needs to be.

> | `aria-label="[landmark description]"` | `.pf-c-nav` |  Describes `<nav>` landmark. |

Author: I don't have particularly strong opinions about it, mostly just whatever we think makes the most sense and looks the cleanest. I don't mind taking it off for some that are more obvious, like labeling, and add anything missing to the description while keeping the content of some of them. For example, I don't think it makes sense to take off `aria-current="page"` because I don't think the average user would have any idea what to fill that with.

> | `aria-label="[landmark description]"` | `.pf-c-nav` |  Describes `<nav>` landmark. |

Reviewer: If anything we could see if content design would have any recommendations.

@srambach to clarify, you prefer the syntax of `aria-label="[landmark description]"` for the first column right? Basically having a placeholder value for any attribute that should have something passed in (vs an attribute like `hidden`)?

> | `aria-label="[landmark description]"` | `.pf-c-nav` |  Describes `<nav>` landmark. |

Reviewer 2: I like having a placeholder, yes.

  </ListItem>
  <ListItem>
    <Checkbox id="navigation-a11y-checkbox-2" label="Standard keyboard navigation can be used to navigate between focusable elements within the navigation." description={<span><kbd>Tab</kbd> navigates to the next focusable element, and <kbd>Shift</kbd> + <kbd>Tab</kbd> navigates to the previous focusable element. Expandable items can be opened with <kbd>Enter</kbd> or <kbd>Space</kbd>.  When the navigation is collapsed, you should not be able to navigate to any of the elements.</span>} />
  </ListItem>
  <ListItem>
    <Checkbox id="navigation-a11y-checkbox-3" label={<span>For expandable or grouped navigation, an accessible name via the <code class="ws-code">title</code> or <code class="ws-code">aria-label</code> is used.</span>} description="This will differentiate each navigation when announced to users of assistive technologies when navigating through a page or a screen reader's rotor menu." />
  </ListItem>
</List>

## React customization

> Various React props have been provided for more fine-tuned control over accessibility.

Reviewer: ```suggestion
The following React props have been provided for more fine-tuned control over accessibility.
```

| Prop | Applied to | Reason | 
|---|---|---|

> | `aria-label` | `Nav` | 	Adds an accessible label for the nav when there are multiple navs on the page |
> | `ariaLeftScroll` | `NavList` | Adds an aria-label for the left scroll button |
> | `ariaRightScroll` | `NavList` | Adds an aria-label for the right scroll button |
> | `id` | `NavGroup` | Used as an identifier to use when there is an `aria-labelledby` on a section element |
> | `title` | `NavGroup` | Adds a title shown for the group |
> | `isActive` | `NavItem` |  Will add the `aria-current` attribute on a nav item when active |
> | `isExpanded` | `NavExpandable` | Adds the `aria-expanded` attribute when is expanded and `hidden` attribute when collapsed |
> | `title` | `NavExpandable` | Adds a title shown for the expandable list |
> | `buttonProps` | `NavExpandable` | Adds any additional props added to the NavExpandable `<button>`|
> | `srText` | `NavExpandable` | If defined, screen readers will read this text instead of the list title |

Reviewer: ```suggestion
| `aria-label="[text describing the nav component]"` | `Nav` | 	Adds an accessible label for the nav when there are multiple navs on the page |
| `ariaLeftScroll="[text describing the left scroll button]"` | `NavList` | Adds an aria-label for the left scroll button |
| `ariaRightScroll="[text describing the right scroll button]"` | `NavList` | Adds an aria-label for the right scroll button |
| `id` | `NavGroup` | Used as an identifier to use when there is an `aria-labelledby` on a section element |
| `title="[text label for a nav group]"` | `NavGroup` | Adds a title shown for the group |
| `isActive` | `NavItem` |  Will add the `aria-current` attribute on a nav item when active |
| `isExpanded` | `NavExpandable` | Adds the `aria-expanded` attribute when is expanded and `hidden` attribute when collapsed |
| `title="[text label for an expandable nav]"` | `NavExpandable` | Adds a title shown for the expandable list |
| `buttonProps={[an object containing button props]}` | `NavExpandable` | Adds any additional props added to the NavExpandable `<button>`|
| `srText="[text describing an expandable nav]"` | `NavExpandable` | If defined, screen readers will read this text instead of the list title |
```

### Expandable toggle props

When a navigation has expandable content, `buttonProps` must be passed into the `NavExpandable` sub-component with an object as its value. This object should usually be passed in with a structure similar to the following:

```noLive
  buttonProps=({
    id: ‘buttonId’, 
    ‘aria-label’: 'Toggle label', 
    ‘aria-labelledby’: 'titleId buttonId', 
    ‘aria-expanded’: this.state.isExpanded
  )}
  ```

> ## HTML/CSS customization
> 
> | Attribute | Applied to | Reason |
> | -- | -- | -- |
> | `aria-label` | `.pf-c-nav` |  Describes `<nav>` landmark. |
> | `aria-label` | `.pf-c-nav__section` |  Describes a nav `<section>`, where a `.pf-c-nav__section-title` is not present. |
> | `aria-labelledby` | `.pf-c-nav__subnav` |  Gives the subnav `<section>` landmark an accessible name by referring to the element that provides the subnav `<section>` landmark title. The `aria-labelledby` attribute should be passed in with a value of the label's id attribute.|

Reviewer: ```suggestion
## HTML/CSS customization

The following HTML attributes and PatternFly classes can be used for more fine-tuned control over accessibility.

| Attribute or class | Applied to | Reason |
| -- | -- | -- |
| `aria-label="[text describing the nav component]"` | `.pf-c-nav` |  Describes `<nav>` landmark. |
| `aria-label="[text describing a nav section]"` | `.pf-c-nav__section` |  Describes a nav `<section>`, where a `.pf-c-nav__section-title` is not present. |
| `aria-labelledby="[id of the element that labels the subnav]"` | `.pf-c-nav__subnav` |  Gives the subnav `<section>` landmark an accessible name by referring to the element that provides the subnav `<section>` landmark title. The `aria-labelledby` attribute should be passed in with a value of the label's id attribute.|
```

| `aria-expanded="false"` | `.pf-c-nav__link` |  Indicates that subnav section is hidden. |
| `aria-expanded="true"` | `.pf-c-nav__link` |  Indicates that subnav section is visible. |
| `hidden` | `.pf-c-nav__subnav` |  Indicates that the subnav section is hidden so that it isn't visible in the UI and isn't accessed by assistive technologies. |
| `disabled` | `.pf-c-nav__scroll-button` | Indicates that a scroll button is disabled, when at the first or last item of a list. **Required when disabled** |
| `aria-current="page"` | `.pf-c-nav__link` |  Indicates the current page link. Can only occur once on page. |
| `aria-haspopup="true"` | `.pf-c-nav__link` | Declares that a nav item has a submenu. |

## Additional considerations

> Consumers must ensure they take any additional considerations when customizing a progress component, using it in a way not described or recommended by PatternFly, or in various other specific use-cases not outlined elsewhere on this page.

Reviewer 2: Is this a paste error?

Author: Yes it is 😂
